### PR TITLE
syncthing: fix cross

### PR DESCRIPTION
**Pull request by yu-re-ka**
###### Description of changes

https://github.com/NixOS/nixpkgs/issues/216480

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/syncthing/default.nix [hunk 1]
```diff
@@ -22,7 +48,9 @@ let
 
       buildPhase = ''
         runHook preBuild
-        go run build.go -no-upgrade -version v${version} build ${target}
```
Comment (NickCao) [inline pkgs/applications/networking/syncthing/default.nix:25]: the go compiler (just like clang) supports all the targets in a single binary. So we can set `GOOS="${pkgsBuildBuild.go.GOOS}" GOARCH="${pkgsBuildBuild.go.GOARCH}"` and everything just work.

Comment (yu-re-ka) [reply pkgs/applications/networking/syncthing/default.nix:25]: Can not confirm:


```
syncthing-armv7l-unknown-linux-gnueabihf> building
syncthing-armv7l-unknown-linux-gnueabihf> ++ export GOOS=linux GOARCH=amd64
syncthing-armv7l-unknown-linux-gnueabihf> ++ GOOS=linux
syncthing-armv7l-unknown-linux-gnueabihf> ++ GOARCH=amd64
syncthing-armv7l-unknown-linux-gnueabihf> ++ go build build.go
syncthing-armv7l-unknown-linux-gnueabihf> # runtime/cgo
syncthing-armv7l-unknown-linux-gnueabihf> armv7l-unknown-linux-gnueabihf-gcc: error: unrecognized command-line option '-m64'
```

It uses the wrong linker

Comment (NickCao) [reply pkgs/applications/networking/syncthing/default.nix:25]: Try additionally setting `CC="${pkgsBuildBuild.stdenv.cc}/bin/cc`

**Comment by yu-re-ka** [comment]
This would work:

```diff
diff --git a/pkgs/applications/networking/syncthing/default.nix b/pkgs/applications/networking/syncthing/default.nix
index 3b171070a57..a01d7afcf76 100644
--- a/pkgs/applications/networking/syncthing/default.nix
+++ b/pkgs/applications/networking/syncthing/default.nix
@@ -1,4 +1,4 @@
-{ buildGoModule, stdenv, lib, procps, fetchFromGitHub, nixosTests }:
+{ pkgsBuildBuild, go, buildGoModule, stdenv, lib, procps, fetchFromGitHub, nixosTests }:
 
 let
   common = { stname, target, postInstall ? "" }:
@@ -15,6 +15,7 @@ let
 
       vendorHash = "sha256-q63iaRxJRvPY0Np20O6JmdMEjSg/kxRneBfs8fRTwXk=";
 
+
       doCheck = false;
 
       BUILD_USER = "nix";
@@ -22,7 +23,12 @@ let
 
       buildPhase = ''
         runHook preBuild
-        go run build.go -no-upgrade -version v${version} build ${target}
+        (
+          export GOOS="${pkgsBuildBuild.go.GOOS}" GOARCH="${pkgsBuildBuild.go.GOARCH}" CC=$CC_FOR_BUILD
+          go build build.go
+          go generate github.com/syncthing/syncthing/lib/api/auto github.com/syncthing/syncthing/cmd/strelaypoolsrv/auto
+        )
+        ./build -goos ${go.GOOS} -goarch ${go.GOARCH} -no-upgrade -version v${version} build ${target}
         runHook postBuild
       '';
 ```

would you prefer that @NickCao ?

**Comment by NickCao** [comment]
Of course yes, this saves us fiddling with an additional derivation (with the downside of looking a bit messy). Actually this cross compiling issue exists for all go packages using `go generate`, which should really be fixed upstream.

**Comment by Skallwar** [comment]
> ```diff
>  export GOOS="${pkgsBuildB
> ```

A note on doing `go build build.go` then `./build`, on my machine I got a `Permission denied` error when trying to run `./build`. That's why I created another binary on my comments in the issue. I'm not familiar with nix and go, so it might be me doing some stuff wrong

**Comment by yu-re-ka** [comment]
I think we can safely assume that the freshly built binary is executable

**Comment by Skallwar** [comment]
> I think we can safely assume that the freshly built binary is executable

That's what I assumed but:
- Permission looked good (executable)
- Readelf confirmed that the binary was x86 and not aarch64
- Owner and group same as every other file in the folder (used ls -lah)
But still I got a permission denied.

Anyway, if it works for you then I guess I'm happy :D

I noticed the use of parenthesis. What does it means ?

**Comment by NickCao** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.syncthing

**Review by NickCao — APPROVED** [review]

**Comment by NickCao** [comment]
Waiting a while before merging, also cc @zowoq to see if there are better/generalized solutions.

**Comment by zowoq** [comment]
Nothing I can suggest offhand but I'll make a note that we should handle this somehow in `buildGo*`.

**Pull request merged**